File: Stocks/frontend/src/types/stock.ts
export interface StockSearchResult {
  symbol: string
  name: string
  exchange: string
  nameCn?: string
}

export interface InstitutionalRecord {
  date: string
  foreignBuy: number
  foreignSell: number
  foreignNet: number
  itBuy: number
  itSell: number
  itNet: number
  dealerBuy: number
  dealerSell: number
  dealerNet: number
  totalBuy: number
  totalSell: number
  totalNet: number
}

export interface InstitutionalData {
  symbol: string
  data: InstitutionalRecord[]
}

export interface StockInfo {
  symbol: string
  name: string
  nameCn?: string
  nameEn?: string
  currentPrice: number
  previousClose: number
  open: number
  dayHigh: number
  dayLow: number
  change: number
  changePercent: number
  marketCap: number
  volume: number
  avgVolume: number
  peRatio: number | null
  forwardPE: number | null
  eps: number | null
  forwardEps: number | null
  earningsDate: string | null
  dividendYield: number | null
  dividendRate: number | null
  exDividendDate: string | null
  payoutRatio: number | null
  fiveYearAvgDividendYield: number | null
  dividendFrequency?: string
  meetingUrl?: string
  roe: number | null
  roa: number | null
  revenue: number | null
  revenuePerShare: number | null
  profitMargin: number | null
  operatingMargin: number | null
  debtToEquity: number | null
  bookValue: number | null
  priceToBook: number | null
  fiftyTwoWeekHigh: number
  fiftyTwoWeekLow: number
  fiftyTwoWeekChange: number | null
  beta: number | null
  sector: string
  industry: string
  country: string
  website: string
  description: string
  employees: number | null
  exchange: string
  currency: string
  logoUrl: string | null
  // Premium / fair value
  premium?: number | null
  fairValue?: number | null
  fairValueMethod?: string | null
  // Company management info (Taiwan stocks)
  chairman?: string | null
  generalManager?: string | null
  spokesperson?: string | null
  spokespersonTitle?: string | null
  deputySpokesperson?: string | null
  establishedDate?: string | null
  listingDate?: string | null
  phone?: string | null
  companyAddress?: string | null
  capital?: number | null
  shareTransferAgency?: string | null
  auditorFirm?: string | null
  auditor1?: string | null
  auditor2?: string | null
  fax?: string | null
  companyEmail?: string | null
  isETF?: boolean
  // ETF-specific fields
  ytdReturn?: number | null
  totalAssets?: number | null
  navPrice?: number | null
  threeYearAverageReturn?: number | null
  fiveYearAverageReturn?: number | null
  annualReportExpenseRatio?: number | null
  fundFamily?: string | null
  category?: string | null
}

export interface ChartDataPoint {
  date: string
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface ChartResponse {
  symbol: string
  period: string
  interval: string
  data: ChartDataPoint[]
}

export interface DividendItem {
  date: string
  amount: number
}

export interface StockDividends {
  symbol: string
  dividends: DividendItem[]
  splits: { date: string; ratio: number }[]
}

export interface FinancialData {
  [statement: string]: {
    [label: string]: { [year: string]: number | string | null }
  }
}

export interface RealtimePrice {
  symbol: string
  price: number
  change: number
  changePercent: number
  timestamp: string
}

export interface MajorHolders {
  [key: string]: string[]
}

export interface InstitutionalHolder {
  Holder: string
  Shares: string
  Date_Reported: string
  pctHeld: string
  Value: string
}

export interface SentimentSignal {
  factor: string
  value: string
  signal: 'bullish' | 'bearish' | 'neutral'
  reason?: string
}

export interface Recommendation {
  buy: number
  hold: number
  period: string
  sell: number
  strongBuy: number
  strongSell: number
  symbol: string
}

export interface SentimentData {
  symbol: string
  overall: 'bullish' | 'bearish' | 'neutral'
  score: number
  bullishCount: number
  bearishCount: number
  signals: SentimentSignal[]
  recommendations: Recommendation[]
  institutionalTrading: Record<string, number | string>
}

export interface EtfNavRecord {
  date: string
  nav: number | null
  price: number | null
  premium: number | null
}

export interface EtfNavData {
  symbol: string
  currentNAV: number | null
  currentPrice: number | null
  premium: number | null
  navPreviousClose: number | null
  history: EtfNavRecord[]
}

export interface EtfHolding {
  symbol: string
  name: string
  weight: number | null
}

export interface EtfHoldingsData {
  symbol: string
  holdings: EtfHolding[]
}

export interface AiConsultResponse {
  answer: string
}

export interface EtfConcentration {
  top1Symbol: string
  top1Name: string
  top1Weight: number
  top5Weight: number
  top10Weight: number
  totalHoldings: number
  herfindahlIndex: number
}

export interface EtfSectorExposure {
  sector: string
  weight: number
}

export interface EtfFeeAnalysis {
  expenseRatio: number | null
  expenseRatioPercent: number | null
  feeRating: string
  categoryAvgExpenseRatio: number
  totalAssets: number | null
  totalAssetsFormatted: string | null
  fundFamily: string | null
  category: string | null
  managementName: string | null
  managementSince: string | null
}

export interface EtfDistributionAnalysis {
  dividendCount12m: number
  averageDividend: number
  totalDividend12m: number
  distributionYield: number | null
  distributionFrequency: string
}

export interface EtfPremiumDiscountAnalysis {
  averagePremium: number | null
  maxPremium: number | null
  minPremium: number | null
  premiumDays: number
  discountDays: number
  neutralDays: number
  premiumRatio: number
  dataPoints: number
  currentPremium: number | null
  currentNAV: number | null
}

export interface EtfPerformanceAnalysis {
  ytdReturn: number | null
  threeYearAverageReturn: number | null
  fiveYearAverageReturn: number | null
}

export interface EtfScoreDetail {
  factor: string
  score: number
  maxScore: number
  detail: string
}

export interface EtfScore {
  totalScore: number
  maxScore: number
  percentage: number
  details: EtfScoreDetail[]
  rating: string
}

export interface EtfAnalysisData {
  symbol: string
  concentration: EtfConcentration
  sectorExposure: EtfSectorExposure[]
  feeAnalysis: EtfFeeAnalysis
  distributionAnalysis: EtfDistributionAnalysis
  premiumDiscountAnalysis: EtfPremiumDiscountAnalysis
  performanceAnalysis: EtfPerformanceAnalysis
  score: EtfScore
}

File: Stocks/flutter/sotcks/lib/models/stock.dart
class StockSearchResult {
  final String symbol;
  final String name;
  final String exchange;
  final String? nameCn;

  StockSearchResult({
    required this.symbol,
    required this.name,
    required this.exchange,
    this.nameCn,
  });

  factory StockSearchResult.fromJson(Map<String, dynamic> json) {
    return StockSearchResult(
      symbol: json['symbol'] ?? '',
      name: json['name'] ?? '',
      exchange: json['exchange'] ?? '',
      nameCn: json['nameCn'],
    );
  }
}

class StockInfo {
  final String symbol;
  final String name;
  final String? nameCn;
  final double currentPrice;
  final double previousClose;
  final double open;
  final double dayHigh;
  final double dayLow;
  final double change;
  final double changePercent;
  final double? marketCap;
  final int volume;
  final double? peRatio;
  final double? forwardPE;
  final double? eps;
  final double? dividendYield;
  final double? dividendRate;
  final String? exDividendDate;
  final double? payoutRatio;
  final double? roe;
  final double? roa;
  final double? revenue;
  final double? profitMargin;
  final double? debtToEquity;
  final double? bookValue;
  final double? priceToBook;
  final double? fiftyTwoWeekHigh;
  final double? fiftyTwoWeekLow;
  final double? beta;
  final String sector;
  final String? description;
  final String exchange;
  final String currency;
  final String? logoUrl;
  final double? avgVolume;
  final double? revenuePerShare;
  final double? fiveYearAvgDividendYield;
  final double? forwardEps;
  final double? operatingMargin;
  final double? fiftyTwoWeekChange;

  StockInfo({
    required this.symbol,
    required this.name,
    this.nameCn,
    required this.currentPrice,
    required this.previousClose,
    required this.open,
    required this.dayHigh,
    required this.dayLow,
    required this.change,
    required this.changePercent,
    this.marketCap,
    required this.volume,
    this.peRatio,
    this.forwardPE,
    this.eps,
    this.dividendYield,
    this.dividendRate,
    this.exDividendDate,
    this.payoutRatio,
    this.roe,
    this.roa,
    this.revenue,
    this.profitMargin,
    this.debtToEquity,
    this.bookValue,
    this.priceToBook,
    this.fiftyTwoWeekHigh,
    this.fiftyTwoWeekLow,
    this.beta,
    required this.sector,
    this.description,
    required this.exchange,
    required this.currency,
    this.logoUrl,
    this.avgVolume,
    this.revenuePerShare,
    this.fiveYearAvgDividendYield,
    this.forwardEps,
    this.operatingMargin,
    this.fiftyTwoWeekChange,
  });

  factory StockInfo.fromJson(Map<String, dynamic> json) {
    double n(String key) => (json[key] ?? 0).toDouble();
    double? opt(String key) => json[key] != null ? (json[key] as num).toDouble() : null;
    String s(String key) => json[key]?.toString() ?? '';
    int i(String key) => (json[key] ?? 0).toInt();

    final cp = n('currentPrice');
    final pc = n('previousClose');
    final c = cp - pc;
    final cpct = pc > 0 ? c / pc * 100 : 0.0;

    return StockInfo(
      symbol: s('symbol'),
      name: s('name'),
      nameCn: json['nameCn']?.toString(),
      currentPrice: cp,
      previousClose: pc,
      open: n('open'),
      dayHigh: n('dayHigh'),
      dayLow: n('dayLow'),
      change: opt('change') ?? c,
      changePercent: opt('changePercent') ?? cpct,
      marketCap: opt('marketCap'),
      volume: i('volume'),
      peRatio: opt('peRatio'),
      forwardPE: opt('forwardPE'),
      eps: opt('eps'),
      dividendYield: opt('dividendYield'),
      dividendRate: opt('dividendRate'),
      exDividendDate: json['exDividendDate']?.toString(),
      payoutRatio: opt('payoutRatio'),
      roe: opt('roe'),
      roa: opt('roa'),
      revenue: opt('revenue'),
      profitMargin: opt('profitMargin'),
      debtToEquity: opt('debtToEquity'),
      bookValue: opt('bookValue'),
      priceToBook: opt('priceToBook'),
      fiftyTwoWeekHigh: opt('fiftyTwoWeekHigh'),
      fiftyTwoWeekLow: opt('fiftyTwoWeekLow'),
      beta: opt('beta'),
      sector: s('sector'),
      description: json['description']?.toString(),
      exchange: s('exchange'),
      currency: s('currency'),
      logoUrl: json['logoUrl']?.toString(),
      avgVolume: opt('avgVolume'),
      revenuePerShare: opt('revenuePerShare'),
      fiveYearAvgDividendYield: opt('fiveYearAvgDividendYield'),
      forwardEps: opt('forwardEps'),
      operatingMargin: opt('operatingMargin'),
      fiftyTwoWeekChange: opt('fiftyTwoWeekChange'),
    );
  }
}

class ChartDataPoint {
  final String date;
  final double close;

  ChartDataPoint({required this.date, required this.close});

  factory ChartDataPoint.fromJson(Map<String, dynamic> json) {
    return ChartDataPoint(
      date: json['date'] ?? '',
      close: (json['close'] ?? 0).toDouble(),
    );
  }
}

class InstitutionalRecord {
  final String date;
  final double foreignBuy;
  final double foreignSell;
  final double foreignNet;
  final double itBuy;
  final double itSell;
  final double itNet;
  final double dealerBuy;
  final double dealerSell;
  final double dealerNet;
  final double totalBuy;
  final double totalSell;
  final double totalNet;

  InstitutionalRecord({
    required this.date,
    required this.foreignBuy,
    required this.foreignSell,
    required this.foreignNet,
    required this.itBuy,
    required this.itSell,
    required this.itNet,
    required this.dealerBuy,
    required this.dealerSell,
    required this.dealerNet,
    required this.totalBuy,
    required this.totalSell,
    required this.totalNet,
  });

  factory InstitutionalRecord.fromJson(Map<String, dynamic> json) {
    double v(String key) => (json[key] ?? 0).toDouble();
    return InstitutionalRecord(
      date: json['date'] ?? '',
      foreignBuy: v('foreignBuy'),
      foreignSell: v('foreignSell'),
      foreignNet: v('foreignNet'),
      itBuy: v('itBuy'),
      itSell: v('itSell'),
      itNet: v('itNet'),
      dealerBuy: v('dealerBuy'),
      dealerSell: v('dealerSell'),
      dealerNet: v('dealerNet'),
      totalBuy: v('totalBuy'),
      totalSell: v('totalSell'),
      totalNet: v('totalNet'),
    );
  }
}
